packages/block-library/src/categories/index.php

```diff
@@ -54,6 +54,12 @@ function render_block_core_categories( $attributes ) {
 		$wrapper_markup = '<ul %1$s>%2$s</ul>';
 		$items_markup   = wp_list_categories( $args );
 		$type           = 'list';
+
+		$p = new WP_HTML_Tag_Processor( $items_markup );
```

Comment:
Wouldn't it be the responsibility of the Query Loop block to intercept all clicks to links instead when the proper setting is enabled? Otherwise, the same change would have to be applied to every possible block that contains link.

Reply:
Indeed. We should add the `data-wp-on-click` directives in the Query block's render callback.

However, in order to find where to add them, we should also add some "markers" here (in the Categories block). 

One solution could be something like adding `data-wp-navigation-link` attributes (which were removed in https://github.com/WordPress/gutenberg/pull/57853):

```php
// Categories block
while ( $p->next_tag( 'a' ) ) {
	$p->set_attribute( 'data-wp-navigation-link', );
}
```

```php 
// Query block
while ( $p->next_tag( 'a' ) ) {
	if ( $p->get_attribute('data-wp-navigation-link') ) {
	  $p->set_attribute( 'data-wp-on--click', 'core/query::actions.navigate' );
	}
}
$items_markup = $p->get_updated_html();

Reply:
Thanks, folks, that's exactly the kind of feedback I was looking for!

@gziolo wrote:
> Wouldn't it be the responsibility of the Query Loop block to intercept all clicks to links instead when the proper setting is enabled? Otherwise, the same change would have to be applied to every possible block that contains link.

@michalczaplinski wrote:
> Indeed. We should add the `data-wp-on-click` directives in the Query Loop block's render callback.
> However, in order to find where to add them, we should also add some "markers" here (in the Categories block).

I like the idea of adding the directives from within the Query block's render callback, as that gives us access to the [information whether enhanced pagination is on or off](https://github.com/WordPress/gutenberg/blob/520fa41016d9e469aee9328724ac78744f257bf1/packages/block-library/src/query/index.php#L19-L22).

However, we need to decide on an overall strategy:
- Doing it wholesale for all links found inside of the Query Loop block, as @gziolo suggested, or
- Applying it only to individual blocks that have opted in, e.g. via some sort of marker (another `data-wp-` attribute), as suggested by @michalczaplinski.

I think both approaches have their tradeoffs: 
- Doing it wholesale might erroneously apply to links that point other parts of the site (that use a different template and would thus break region-based client-side navigation), or even external links. OTOH.
- Requiring opt-in adds some overhead. Furthermore, in terms of _how_ blocks opt into this behavior, I don't really love the idea of using _another_ `data-wp` attribute just to tell WP to replace it with `data-wp-on--click='core/query::actions.navigate'`. (To me, it raises the question what we're gaining from that, vs. having the block set the `data-wp-on--click` attribute itself.)

We might be able to tweak both strategies a bit:
- Wholesale: We could make sure to only add the `data-wp-on--click` handler to links that point to routes that are handled by the same template. (This could be a bit tricky in practice due to WP's template hierarchy; the system would need to know a lot about routes and templates.)
- Opt-in: Instead of setting a marker on each instance of a given block; can we just opt the entire block type in, i.e. set a block attribute or something of the sort?

Reply:
I see that the Query Loop block `provides` `enhancedPagination` context, which is used by the Post Template block. This looks promising to me, as it might allow us to minimize the extra information we introduce and/or communicate. It would still leave the responsibility of setting `data-wp-on--click='core/query::actions.navigate'` attribute with individual blocks, but based on the above, that might be reasonable.

Reply:
Tried it in https://github.com/WordPress/gutenberg/pull/64907/commits/d9dc1b7941fbac39320017625df5d9fb8d5eaa73. I kinda like this 😬

Reply:
I don't have much to add here.

- It's true that click handlers cannot be added to all the clicks within the Query loop because there are links that may point to other parts of the site.
- I don't have a strong opinion on how to include those click handlers, but in my opinion, the way we are currently doing it, passing whether the enhanced pagination is active through block context and injecting the click handlers into each block, seems simple and appropriate for now.
- Sometimes, there will be more complex cases where a click-handler is not enough, and more logic needs to be added, such as with the Search block, so it's important to leave room for custom implementations.
  - https://github.com/WordPress/gutenberg/pull/63147

Reply:
Thank you very much for weighing in, @luisherranz!

I'll go ahead and open this for review then 😊

Reply:
I left more thoughts related to that topic https://github.com/WordPress/gutenberg/pull/64907#issuecomment-2326316791 in response to the https://github.com/WordPress/gutenberg/pull/64907#pullrequestreview-2276122308 from @dmsnell.